File: internal/installer/components.go
package installer

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// Component represents a single installable component parsed from
// a platform installation script.
type Component struct {
	Name string
}

var arrayRegex = regexp.MustCompile(`(?s)component_installation=\((.*?)\)`)

// ParseComponents reads the installation script for the given system
// and extracts the component_installation bash array entries.
// Commented-out entries (lines starting with #) are skipped.
func ParseComponents(dotfilesDir, system string) ([]Component, error) {
	scriptPath := filepath.Join(dotfilesDir, "script", system+"_installation.sh")
	data, err := os.ReadFile(scriptPath)
	if err != nil {
		return nil, fmt.Errorf("read installation script: %w", err)
	}

	matches := arrayRegex.FindSubmatch(data)
	if matches == nil {
		return nil, fmt.Errorf("no component_installation array found in %s", scriptPath)
	}

	var components []Component
	for _, line := range strings.Split(string(matches[1]), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		// Skip fully commented lines
		if strings.HasPrefix(line, "#") {
			continue
		}
		// Strip inline comments
		if idx := strings.Index(line, "#"); idx > 0 {
			line = strings.TrimSpace(line[:idx])
		}
		if line != "" {
			components = append(components, Component{Name: line})
		}
	}

	return components, nil
}

File: crates/otobun/src/installer/components.rs
use anyhow::{Context, Result};
use regex::Regex;
use std::fs;
use std::path::Path;

#[derive(Debug, Clone)]
pub struct Component {
    pub name: String,
}

pub fn parse_components(dotfiles_dir: &Path, system: &str) -> Result<Vec<Component>> {
    let script_path = dotfiles_dir
        .join("script")
        .join(format!("{system}_installation.sh"));

    let content = fs::read_to_string(&script_path)
        .with_context(|| format!("read installation script: {}", script_path.display()))?;

    let re = Regex::new(r"(?s)component_installation=\((.*?)\)").unwrap();
    let captures = re
        .captures(&content)
        .ok_or_else(|| anyhow::anyhow!("no component_installation array found in {}", script_path.display()))?;

    let array_body = &captures[1];
    let mut components = Vec::new();

    for line in array_body.lines() {
        let line = line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        // Strip inline comments
        let line = match line.find('#') {
            Some(idx) if idx > 0 => line[..idx].trim(),
            _ => line,
        };
        if !line.is_empty() {
            components.push(Component {
                name: line.to_string(),
            });
        }
    }

    Ok(components)
}
